### PR TITLE
 [SofaPython] Cleaner Script Controllers

### PR DESCRIPTION
**Pull request by maxime-tournier**
Hello all,

This PR somewhat simplifies the creation and handling of `PythonScriptControllers` from python scenes.

Here's the TL;DR version:

```python
import Sofa    
class Script(Sofa.PythonScriptController): pass

def createScene(node):
    # now you can do
    script = Script(node, 'bar', egg = 'spam')

    # the old behavior still works:
    old = node.createObject('PythonScriptController', filename = '', classname = 'Script')
```

Now for the details:

## Existing approach
The current approach is to use 
```python
node.createObject('PythonScriptController', filename = '', classname = 'Script')
```
which has several drawbacks:
- The returned object does not point to the control object, but to the component wrapper. This makes it complicated to pass data around at scene creation (people generally resort to `global` variables for this purpose).
- If `filename` is non-empty, the corresponding file is (re)loaded, causing all sorts of *interesting* side-effects. If you ever had to struggle with these, you know what I mean.
- If `filename` is empty, the control class has to lie in the `__main__` namespace
- It is quite verbose
- It just feels plain wrong *not* to create python objects directly when called from python

It does have a few advantages though:
- It does not require any `PythonScriptController`-specific code for object creation,
- It enables the creation of `PythonScriptController` instances from XML scenes 

## Current workaround
There is a workaround class `SofaPython.script.Controller` that somewhat fixes some of the issues by implementing `__new__` in a derived Python class, intercepting and returning the control instance and optional initialization arguments, but overall it feels more like a hack.

## This PR

This PR simply provides a clean binding for instance creation. You need to provide a node as a first argument, but otherwise Controllers now act as any other python object. The old behavior still works, should you need it.

```python
    
import Sofa    
class Script(Sofa.PythonScriptController):

    def __init__(self, node, *args, **kwargs):
        self.bacon = kwargs.pop('egg', 'spam')

    def onBeginAnimationStep(self, dt):
        print(self.bacon)

        
def createScene(node):
    script = Script(node, egg = 'spam')
    

```

Creating instances directly no longer calls `onLoaded` nor `createGraph`:

- nothing is actually loaded
- code in `createGraph` really belongs to `__init__` anyways

This behavior can be changed of course, maybe a warning can/should be emitted if the instance creation detects a derived implementation. Comments welcome :)

The PR also removes heavy copypasta in some files, removes unneeded `extern "C"` linkage and a few other cosmetic changes.

# Changelog

- `PythonScriptController` and derived classes are now instantiable directly in python. The constructor requires a `Node` as first parameter.
- `onLoaded` and `createGraph` entry points are no longer called when classes are instantiated directly from python. Corresponding code belongs to the derived class constructor.
- creating instances "the old way" through `node.createGraph` has unchanged behavior
- the workaround module `SofaPython.script` now features a deprecation message, together with a derived class easing transition (calls `createGraph` on construction, warns if `onLoaded` is defined)
- updated `ControllerVariable.py` example 
- some minor cleanups
______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by matthieu-nesme** [comment]
Super great contrib :) The as-clean-as-possible way to implement a controller.

From a quick review, everything looks good to me.

A few suggestions:
1- what about adding a python-based layer (inherited class) that would, in its `__init__`, set the Data member (like the name) from the kwargs + call an eventual 'createGraph' function + print a warning if 'onLoaded' exists ;)
2- check that everything is ok with the destruction and the reference counter (maybe it is, I am not yet sure)
3- update the ControllerVariable.py example
4- mark the SofaPython.script.Controller as deprecated

**Comment by maxime-tournier** [comment]
Thank you :)
1. totally doable, and would in fact be the preferred way of ensuring smooth transition rather than complicating the c++ side
2. I suspect my proposal has a memleak due to a GC cycle: 
- the python instance wraps and holds a ref to the c++ object, which in turns holds a ref to the instance.
- most likely we want the c++ object to hold a weakref to the python object, but i'm not very familiar with them (yet)
- the previous approach had it too, if I'm not mistaken: `wrapper object -> c++ object -> control object -> c++ object` so by fixing it we kill two birds with one stone
3. will do
4. how?

**Comment by matthieu-nesme** [comment]
2- linked with issue #201 (the famous one)
4- Sofa.msg_deprecated('SofaPython','script.Controller is deprecated, have a look to blablabla')

**Comment by maxime-tournier** [comment]
3 & 4 done. 

1: not sure how we should name the class, if ever
2: harder. i have a toy [implementation](https://github.com/maxime-tournier/cpp/tree/master/pybind) to play with.

briefly: we can put gc flags on the python type, implement `traverse` and `Py_VISIT` the pointer when the c++ object becomes unreachable (c++ refcount == 1)

Then the GC will correctly detect cycles as unreachable objects (using debugging in the `gc` module), but won't delete them, probably since they have a non-trivial `dealloc` slot.

Maybe this will be fixed after we switch to >= 3.4, since finalizers for cycles will be called.

**Comment by maxime-tournier** [comment]
Also, I'm seeing this pattern over and over again:
```c++
    if (!PyArg_ParseTuple(args, "s",&filename)) {
        PyErr_BadArgument();
        return NULL;
    }
```
Please don't do that since it replaces the slightly helpful error message set by `PyArg_ParseTuple` by a generic one. Simply return `NULL` instead:

```c++
    if (!PyArg_ParseTuple(args, "s",&filename)) {
        return NULL;
    }
```

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
Great contribution, 

I fully support any step toward smoother integration of Sofa with Python and this PR is one.
I also like the cleaning & factoring work with stuff like get_base function. 

Really minor details: 
In SOFA we have plenty of TODOs forgotten since years that no one is responsible for, without any idea when this was added and when this will be done. To avoid that I suggest replacing things like
```cpp
 // TODO FIXME self is commented
```
With one with more details
```cpp
//TODO(mtournier 2017-05-30) Self, is commented there is possibly a problem (if not done in two years, you can remove this comment)
``` 
Or
```cpp
//DEFECT(mtournier 2017-05-30) Self, is commented there is a problem when...
``` 
To help general code maintenance.

**Comment by matthieu-nesme** [comment]
I would even suggest not to add TODO but to fix them!!

**Comment by matthieu-nesme** [comment]
We decided to let the point 2 for later (another PR).

We are waiting for the point 1 before merging. If you need a name, I suggest SofaPython.Controller (but anything is ok!)

**Comment by matthieu-nesme** [comment]
@maxime-tournier Thanks for your responsiveness ;)

Mergeable to me.

**Comment by maxime-tournier** [comment]
(bump)

**Comment by damienmarchal** [comment]
[ci-build]

I will merge the PR if it compile nicely. 

@maxime-tournier can you add a synthetic CHANGELOG section at the end of the description of this PR so that @hugtalbot  & @guparan can easily update the CHANGELOG.md file in in sofa.

**Review by matthieu-nesme — APPROVED** [review]

**Comment by damienmarchal** [comment]
I just noticed this PR is adding new feature without providing associated tests.
It would be great to have two tests, one validating the old behavior and one validating the new one.

**Comment by damienmarchal** [comment]
[ci-build]

**Pull request merged**
